### PR TITLE
[FLINK-21628][python] Support Python UDAF in Tumbling Window

### PR DESCRIPTION
**Pull request by HuangXingBo**
## What is the purpose of the change

*This pull request will support Python UDAF in Tumbling Window*

## Brief change log

  - *Add `window_aggregate.py`, `window_process_function.py`, `window_context.py` and `window_trigger.py`*

## Verifying this change

This change added tests and can be verified as follows:

  - *Added integration tests `test_tumbling_group_window_over_time` and `test_tumbling_group_window_over_count` in `test_udaf.py`*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-python/pyflink/fn_execution/window_trigger.py [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+from abc import abstractmethod
+from typing import Generic
+
+from pyflink.common.typeinfo import Types
+from pyflink.datastream.state import ValueStateDescriptor
+from pyflink.fn_execution.window import TimeWindow, CountWindow
+from pyflink.fn_execution.window_context import TriggerContext, W
+
+
+class Trigger(Generic[W]):
```
Comment (WeiZhong94) [inline flink-python/pyflink/fn_execution/window_trigger.py:27]: ditto

File: flink-python/src/main/java/org/apache/flink/streaming/api/runners/python/beam/BeamPythonFunctionRunner.java [hunk 2]
```diff
@@ -198,6 +198,7 @@ public BeamPythonFunctionRunner(
             FlinkMetricContainer flinkMetricContainer,
             @Nullable KeyedStateBackend keyedStateBackend,
             @Nullable TypeSerializer keySerializer,
+            TypeSerializer namespaceSerializer,
```
Comment (WeiZhong94) [inline flink-python/src/main/java/org/apache/flink/streaming/api/runners/python/beam/BeamPythonFunctionRunner.java:201]: add Nullable

**Review by WeiZhong94 — COMMENTED** [review]
@HuangXingBo Thanks for your PR. I have left some comments for you.

File: flink-python/pyflink/fn_execution/coder_impl_fast.pyx [hunk 3]
```diff
@@ -107,8 +118,11 @@ cdef class AggregateFunctionRowCoderImpl(FlattenRowCoderImpl):
                 self._encode_internal_row(value, output_stream)
         else:
             for result in results:
-                for value in result:
-                    self._encode_internal_row(value, output_stream)
+                if isinstance(result[0], InternalRow):
```
[the remark this answers is not in the document]
Comment (HuangXingBo) [reply flink-python/pyflink/fn_execution/coder_impl_fast.pyx:121]: Yes. Make sense.

File: flink-python/pyflink/fn_execution/operations.py [hunk 4]
```diff
@@ -406,6 +420,121 @@ def create_process_function(self, user_defined_aggs, input_extractors, filter_ar
...
+                for timer in timers:
+                    timer_operand_type = timer[0]  # type: TimerOperandType
+                    internal_timer = timer[1]  # type: InternalTimer
+                    window = internal_timer.get_namespace()
+                    key = internal_timer.get_key()
+                    timestamp = internal_timer.get_timestamp()
+                    if self._is_time_window:
+                        timer_data = \
+                            [TRIGGER_TIMER, None,
+                             [timer_operand_type.value, key, timestamp, window.start, window.end]]
```
[the remark this answers is not in the document]
Comment (HuangXingBo) [reply flink-python/pyflink/fn_execution/operations.py:493]: Yes. Make sense

**Comment by HuangXingBo** [comment]
@WeiZhong94 Thanks a lot for the review. I have addressed the comments at the latest commit.

**Review by WeiZhong94 — APPROVED** [review]

**Pull request closed without merging**